pkg/ingester/activeseries/model/matchers.go

```diff
@@ -3,28 +3,147 @@
 package activeseriesmodel
 
 import (
-	"sort"
+	"cmp"
+	"maps"
+	"slices"
 
 	amlabels "github.com/prometheus/alertmanager/pkg/labels"
 	"github.com/prometheus/prometheus/model/labels"
 )
 
 func NewMatchers(matchersConfig CustomTrackersConfig) *Matchers {
-	asm := &Matchers{cfg: matchersConfig}
-	for name, matchers := range matchersConfig.config {
-		asm.matchers = append(asm.matchers, matchers)
-		asm.names = append(asm.names, name)
+	names := make([]string, 0, len(matchersConfig.config))
+	unoptimized := make([]indexedMatchers, 0, len(matchersConfig.config))
+
+	// singleLabelMatchers is a map of label names to matchers that match on that single label.
+	// This is an optimization under the assumption that there are more of these than labels in a series,
+	// so we don't have to lookup the label value every time.
+	// This does not include matchers that match on empty labels.
+	singleLabelMatchers := make(map[string][]indexedMatchers)
+
+	// matchersForExactValue is a list of matchers that have at least one SetMatcher or exact value matcher.
+	matchersForExactValue := make([]indexedMatchers, 0, len(matchersConfig.config))
+	// As we scan them, keep track of the amount of different values matched per label name.
+	valuesMatchedPerLabelName := map[string]map[string]struct{}{}
+	for name, ms := range matchersConfig.config {
+		index := len(names)
+		names = append(names, name)
+
+		if len(ms) == 1 && !ms[0].Matches("") {
+			// This is a single matcher that matches on a single label.
+			// We can optimize it by treating it as a single label matcher.
+			singleLabelMatchers[ms[0].Name] = append(singleLabelMatchers[ms[0].Name], indexedMatchers{labelsMatchers: ms, index: index})
+			continue
+		}
+
+		optimized := false
+		for _, m := range ms {
+			if m.Type == labels.MatchEqual && m.Value != "" {
+				matchersForExactValue = append(matchersForExactValue, indexedMatchers{labelsMatchers: ms, index: index})
+				if valuesMatchedPerLabelName[m.Name] == nil {
+					valuesMatchedPerLabelName[m.Name] = make(map[string]struct{})
+				}
+				valuesMatchedPerLabelName[m.Name][m.Value] = struct{}{}
+				optimized = true
+				break
+			} else if sm := m.SetMatches(); len(sm) > 0 {
+				matchersForExactValue = append(matchersForExactValue, indexedMatchers{labelsMatchers: ms, index: index})
+				if valuesMatchedPerLabelName[m.Name] == nil {
+					valuesMatchedPerLabelName[m.Name] = make(map[string]struct{})
+				}
+				for _, v := range sm {
+					valuesMatchedPerLabelName[m.Name][v] = struct{}{}
+				}
+				optimized = true
+				break
+			}
+		}
+		if !optimized {
+			unoptimized = append(unoptimized, indexedMatchers{labelsMatchers: ms, index: index})
+		}
+	}
+
+	// Get label names and sort them by popularity (desc).
+	// We select labels that have the most different values matched first to make the decision tree higher and lower.
+	labelNames := slices.Collect(maps.Keys(valuesMatchedPerLabelName))
```

Comment:
Those two statements could be unified:

```suggestion
	labelNames := slices.SortedFunc(
		maps.Keys(valuesMatchedPerLabelName),
		func(a, b string) int {
			return cmp.Compare(len(valuesMatchedPerLabelName[b]), len(valuesMatchedPerLabelName[a]))
		},
	)
```